packages/sync/src/firehose/index.ts

```diff
@@ -0,0 +1,342 @@
+import { createDeferrable, Deferrable, wait } from '@atproto/common'
+import {
+  IdResolver,
+  parseToAtprotoDocument,
+  DidDocument,
+} from '@atproto/identity'
+import {
+  cborToLexRecord,
+  formatDataKey,
+  parseDataKey,
+  readCar,
+  RepoVerificationError,
+  verifyProofs,
+} from '@atproto/repo'
+import { AtUri } from '@atproto/syntax'
+import { Subscription } from '@atproto/xrpc-server'
+import {
+  type Account,
+  type Commit,
+  type Identity,
+  type RepoEvent,
+  RepoOp,
+  isAccount,
+  isCommit,
+  isIdentity,
+  isValidRepoEvent,
+} from './lexicons'
+import {
+  Event,
+  CommitMeta,
+  CommitEvt,
+  AccountEvt,
+  AccountStatus,
+  IdentityEvt,
+} from '../events'
+import { CID } from 'multiformats/cid'
+import { SyncQueue } from '../queue'
+import { didAndSeqForEvt } from '../util'
+
+export type FirehoseOptions = {
+  idResolver: IdResolver
+
+  handleEvt: (evt: Event) => Promise<void>
+  onError: (err: Error) => void
+  getCursor?: () => Promise<number | undefined>
+
+  syncQueue?: SyncQueue // sync queue getCursor takes precedence over `getCursor`
+
+  service?: string
+  subscriptionReconnectDelay?: number
+
+  unauthenticatedCommits?: boolean
+  unauthenticatedHandles?: boolean
+
+  filterCollections?: string[]
+  excludeIdentity?: boolean
+  excludeAccount?: boolean
+  excludeCommit?: boolean
+}
+
+export class Firehose {
+  private sub: Subscription<RepoEvent>
+  private abortController: AbortController
+  private destoryDefer: Deferrable
+
+  constructor(public opts: FirehoseOptions) {
+    this.destoryDefer = createDeferrable()
+    this.abortController = new AbortController()
+    this.sub = new Subscription({
+      service: opts.service ?? 'wss://bsky.network',
+      method: 'com.atproto.sync.subscribeRepos',
+      signal: this.abortController.signal,
+      getParams: async () => {
+        if (this.opts.syncQueue) {
+          return { cursor: this.opts.syncQueue.cursor }
+        } else if (opts.getCursor) {
+          const cursor = await opts.getCursor()
+          return { cursor }
+        }
+        return undefined
+      },
+      validate: (value: unknown) => {
+        try {
+          return isValidRepoEvent(value)
+        } catch (err) {
+          this.opts?.onError(new FirehoseValidationError(err, value))
```

Comment:
I think this is always going to exist.
```suggestion
          this.opts.onError(new FirehoseValidationError(err, value))
```